Set compilers to C++14 and C11

kjbracey opened the pull request:
### Description

* ARMC6 and GCC are set to C++14 and C11.
* ARMC5 is set to C++11 and C99, as it's the highest it can offer.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [X] Breaking change

### Release Notes

* GCC and ARM toolchain profiles now select C++14 and C11, matching IAR, and these are the tested profiles. Codebase in this release should still work if profiles are set to C++98 and C99, although this is no longer tested.

### Migration notes

As the default profiles now select C++14 and C11 for GCC and ARM toolchains, some applications may fail to compile because they use constructs that were valid in C++98, but not in C++14.

IAR 8 has always operated in C++14/C11 mode, so there are no new changes for IAR users who switched to IAR 8 for Mbed OS 5.12, but these notes apply for users migrating from older Mbed OS versions that used IAR 7.

#### Common compatibility issues

* A space is required when a macro follows a string literal, for example in C99-style `printf` formats
```C++
    uint32_t val;
    printf("val = %"PRIu32, val); // Not valid in C++11
    printf("val = %" PRIu32, val); // OK
```
Without the space, C++11 interprets it as being a request for a user-defined "PRIu32-type" string literal.

* Initializer lists cannot have implicit narrowing conversions:
```C++
    uint32_t x;
    uint8_t array1[] = { x }; // Not valid in C++11
    uint8_t array2[] = { 0xffff }; // Not valid in C++11
    uint8_t array3[] = { x & 0xff }; // Not valid in C++11
    uint8_t array4[] = { (uint8_t) x }; // OK
    uint8_t array5[] = { static_cast<uint8_t>(x) }; // OK
    uint8_t array6[] = { 0xffff & 0xff }; // OK (because it's a compile-time constant that fits)
```
These changes should be easy to make to existing codebases. A guide to other possible breakages in C++11 can be found at https://stackoverflow.com/questions/6399615/what-breaking-changes-are-introduced-in-c11. C++14 and C11 cause few extra issues.

#### Future compatibility issues
* The `register` keyword is deprecated in C++14, and is removed in C++17. Some compilers issue warnings for `register` use in C++14, but this has been temporarily suppressed due to the prevalence of the keyword in target code. C++ code should be updated to remove the keyword as soon as possible - the warning will be reactivated once Mbed OS itself no longer triggers it.

#### Fallback

Mbed OS 5.13 releases should still work if compiled as C++98/C99, although this is not tested - if you have serious application compatibility issues, you should be able to switch the build profile back to the earlier language version as a temporary measure. This is likely to no longer be the case for Mbed OS 5.14.

#### ARM Compiler 5

Note also that ARM Compiler 5 has limited C++11 support and no C++14 or C11 support, and ARM Compiler 5 is no longer tested or officially supported. Nevertheless, ARMC5 builds have not been deliberately broken; ARMC5 build profiles select its C++11 mode in an attempt to match the other toolchains as much as possible, but the limited support may itself cause issues - attempts to check `__cplusplus >= 201103` may activate code that ARMC5 cannot actually compile. Like the other toolchains, the profile can be switched back to C++98 if necessary for now.

Continued ARM C 5 support will limit the adoption of C++11 and C++14 features in the Mbed OS codebase, so it is quite possible that ARM Compiler 5 builds may stop working altogether in an upcoming release.

kjbracey committed:
Set compilers to C++14 and C11

* ARMC6 and GCC are set to C++14 and C11.
* ARMC5 is set to C++11 and C99, as it's the highest it can offer.

bridadan commented on tools/export/uvision/__init__.py:
```diff
@@ -183,7 +183,7 @@ def format_flags(self):
         )
         in_template = set(
-            ["--no_vla", "--cpp", "--c99", "-MMD"] + config_option
+            ["--no_vla", "--cpp", "--cpp11", "--c99", "-MMD"] + config_option
```
Are the `--cpp` and `--cpp11` flags exclusive in this case?
```suggestion
            ["--no_vla", "--cpp11", "--c99", "-MMD"] + config_option
```

kjbracey replied:
My reading of that was that that was just a list of options that would be eligible for passthrough in the code below. Can you check?

In which case both `--cpp` and `--cpp11` want this behaviour, in case anyone wants to switch profile back to old version.

kjbracey replied:
> that was that that was

PS Don't report me to tech authors.

bridadan replied:
Oh gosh you're totally right, good eye! In my defense, I have not finished my first cup of coffee this morning 😄

bridadan reviewed (commented):
Very very exciting!!!

Do you know what this does to code size?

bridadan reviewed (approved):
Very nice work!

deepikabhavnani commented on tools/export/coide/arch_max.coproj.tmpl:
```diff
@@ -6,7 +6,7 @@
         <Option name="OptimizationLevel" value="4"/>
         <Option name="UseFPU" value="0"/>
-        <Option name="UserEditCompiler" value="-fno-common; -fmessage-length=0; -Wall; -fno-strict-aliasing; -fno-rtti; -fno-exceptions; -ffunction-sections; -fdata-sections; -std=gnu++98"/>
+        <Option name="UserEditCompiler" value="-fno-common; -fmessage-length=0; -Wall; -fno-strict-aliasing; -fno-rtti; -fno-exceptions; -ffunction-sections; -fdata-sections; -std=gnu++14"/>
```
Question: How do we differentiate Armc5 / Armc6 profiles in exporters?

kjbracey replied:
I don't know, but not sure I understand the question. This template must be for GCC or ARMC6, as `-std` is an option only for them.

There may be further exporter changes required for toolchains where the standard is specified via an XML option, which may not be visible on my search.

deepikabhavnani replied:
My bad, I thought this was default template for uVision didn;t see `coide`

bridadan replied:
Deepika's right though, we'll need to update the uvision template a bit more to play nicely with the C++ and C version. The language versions are selected by the IDE GUI.

bulislaw reviewed (approved):
Approved, if it doesn't inflate the size (please check couple of examples manually, as I can't see the code size checks run in the PR). Also that should go for 5.13 not a patch release.

kjbracey committed:
µVision export: Add ARMC6 `-std` handling

This is limited to ARMC6 because as of µVision V5.27 you can't set C++11
for ARMC5.

Also current µVision does not support gnu++14. We should be able to get
is as `<default>`, as it is the default for ARM Compiler 6.10-6.12,
but this option does not work as documented and actually requests
gnu++89 explicitly. So gnu++14 is mapped to gnu++11.

bridadan commented on tools/export/uvision/uvision.tmpl:
```diff
@@ -369,8 +369,8 @@
-            <v6Lang>4</v6Lang>
-            <v6LangP>2</v6LangP>
+            <v6Lang>6</v6Lang>
+            <v6LangP>4</v6LangP>
```
The uvision template change has to be conditional because the online compiler needs to be able to export older projects that cannot use the newer language standards. The online compiler will use the latest copy of the tools but switch the profiles accordingly so the correct language version is used.

kjbracey replied:
Reworked to parse the `-std` option. And I realised that we can actually work around uVision not knowing about `gnu++14` - it's armclang's default and we can ask for `<default>`.

kjbracey replied:
Except that workaround doesn't seem to work. Contrary to uVision documentation, if you put `<default>` in, it explicitly selects `-std=gnu11` for C and `-std=gnu++89` for C++.

(Those were the defaults for Compiler 6.9, I guess, but the documentation clearly says that it will use the compiler default by not passing an option).

bridadan reviewed (commented):
> Second commit - ("Add ARMC6 -std handling") tries to map `std` to GUI, but ends up turning C++14 into C++11.
> So what I was suggesting was delete the third commit and keep the second to use the GUI and hope/wait for the IDE to be updated. (I've raised an issue).

I think your plan makes sense. We should surface out these control to the GUI boxes instead of overriding them in "Misc Controls".

Could you clarify what you mean about "ends up turning C++14 into C++11"? I can see the "C++14 (community)" option in my drop down box in uvision, are you saying you cannot specify that option through the template?

0xc0170 reviewed (changes requested):
LGTM, just reference to documentation update about this version bump.

mark-edgeworth reviewed (dismissed):
This changeset itself looks good, but the potential impact of such a breaking change is unclear to me. How much testing has been undertaken to ensure that existing code will still compile? What is the impact on users? Are there any codesize or memory implications of this change? Has the change been communicated to the documentation teams?
Is this change necessary right now?
Arm's own workflow rules say that 'breaking changes are always limited to a major version release'

kjbracey committed:
NRF52: C++11 fixes

kjbracey committed:
ARMC6: disable "reserved user-defined literal" warning

Clang warns about reserved user-defined literals by default. This
warning is not terribly helpful; compilers aren't normally in the
habit of warning about use of reserved identifiers. It can interfere
with, for example, deliberate emulation of a future standard
language feature.

The warning was promoted to an error in an mbed client build, due to a
non-C++11 "%s"name occurring in a macro. But the macro itself was never
invoked, so the misinterpretation as C++11 caused no problems other than
this warning. Killing the warning will let that code build on ARMC6.
The code already built on GCC and IAR.

If that macro ever was used, then a separate error about operator ""
name not being defined would be generated, on all 3 toolchains.

kjbracey committed:
FPGA_CI_TEST_SHIELD: C++11 fix
